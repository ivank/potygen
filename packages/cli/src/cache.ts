import { existsSync, readFileSync, writeFileSync, statSync, mkdirSync } from 'fs';
import { dirname } from 'path';
import { createHash } from 'crypto';

export class CachedFileParser<TResult> {
  constructor(public store = new Map<string, TResult>(), public process: (content: string, path: string) => TResult) {}

  public get(path: string): TResult {
    const content = readFileSync(path, 'utf-8');
    const key = createHash('md5').update(content).digest('hex');
    const cachedResult = this.store.get(key);
    if (cachedResult) {
      return cachedResult;
    } else {
      const result = this.process(content, path);
      this.store.set(key, result);
      return result;
    }
  }
}

export class CachedProcessParser<TResult> {
  constructor(public store = new Map<string, TResult>(), public process: (content: string, path: string) => TResult) {}

  public get(path: string): TResult {
    const content = readFileSync(path, 'utf-8');
    const key = createHash('md5').update(content).digest('hex');
    const cachedResult = this.store.get(key);
    if (cachedResult) {
      return cachedResult;
    } else {
      const result = this.process(content, path);
      this.store.set(key, result);
      return result;
    }
  }
}

export class CacheStore {
  public store = new Map<string, number>();

  constructor(public fileName: string, public enabled: boolean = false, public cacheClear: boolean = false) {
    if (this.enabled && !this.cacheClear) {
      if (existsSync(fileName)) {
        this.store = new Map(Object.entries(JSON.parse(readFileSync(fileName, 'utf-8'))));
      }
    }
  }

  public shouldParseFile(path: string) {
    if (!this.enabled) {
      return true;
    } else {
      const mtime = statSync(path).mtime.getTime();
      const currentMtime = this.store.get(path);
      if (currentMtime && mtime === currentMtime) {
        return false;
      }
    }
    return true;
  }

  public cacheFile(path: string) {
    const mtime = statSync(path).mtime.getTime();
    this.store.set(path, mtime);
  }

  public save() {
    if (this.enabled) {
      mkdirSync(dirname(this.fileName), { recursive: true });
      writeFileSync(this.fileName, JSON.stringify(Object.fromEntries(this.store)), 'utf-8');
    }
  }
}
