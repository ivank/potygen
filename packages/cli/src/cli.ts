import { Client } from 'pg';
import { pipeline } from 'stream';
import { promisify } from 'util';
import { SqlRead, QueryLoader } from './traverse';

const asyncPipeline = promisify(pipeline);

const main = async () => {
  const db = new Client({
    database: 'schemes',
    user: 'schemes',
    password: 'dev-pass',
    host: '127.0.0.1',
    port: 5432,
    ssl: false,
  });
  try {
    await db.connect();

    const sqls = new SqlRead('src/**/*.query.ts', '../../../schemes/packages/api');
    const sink = new QueryLoader(
      db,
      '../../../schemes/packages/api',
      '{{root}}/src/queries/__psql-ts__/{{name}}.queries.ts',
    );

    await asyncPipeline(sqls, sink);
  } catch (error) {
    console.log(String(error));
  } finally {
    await db.end();
  }
};

main();
