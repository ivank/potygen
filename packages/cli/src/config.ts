import { Record, Optional, Static, String, Boolean } from 'runtypes';

export const Config = Record({
  files: Optional(String),
  watch: Optional(Boolean),
  root: Optional(String),
  connection: Optional(String),
  template: Optional(String),
  verbose: Optional(Boolean),
  silent: Optional(Boolean),
  typePrefix: Optional(String),
  preload: Optional(Boolean),
  cache: Optional(Boolean),
  cacheFile: Optional(String),
  cacheClear: Optional(Boolean),
});

export type ConfigType = Static<typeof Config>;
export type FullConfigType = Required<ConfigType>;

export const toConfig = (config: unknown): FullConfigType => ({
  files: '**/*.sql',
  cacheFile: '.cache/potygen.cache',
  root: process.cwd(),
  preload: false,
  template: '{{dir}}/{{name}}.queries.ts',
  connection: 'postgres://localhost:5432/db',
  watch: false,
  verbose: false,
  silent: false,
  cache: false,
  cacheClear: false,
  typePrefix: '',
  ...Config.check(config),
});
