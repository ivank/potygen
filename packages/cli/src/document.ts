import { document, printDocument, DocumentContext, Type, withIdentifier } from '@ovotech/ts-compose';
import { TypeNode } from 'typescript';
import { isLoadedArrayType, isLoadedUnionType, isLoadedLiteralType, isLoadedValuesPick } from './guards';
import { LoadedQuery, LoadedValuesPick, LoadedType } from './types';

const toPropertyType = (type: LoadedType | LoadedValuesPick): TypeNode => {
  if (isLoadedValuesPick(type)) {
    return Type.Array(
      Type.TypeLiteral({
        props: type.items.map((item) => Type.Prop({ name: item.name, type: toPropertyType(item.value) })),
      }),
    );
  } else if (isLoadedArrayType(type)) {
    return Type.Array(toPropertyType(type.items));
  } else if (isLoadedUnionType(type)) {
    return Type.Union(type.items.map(toPropertyType));
  } else if (isLoadedLiteralType(type)) {
    return Type.Literal(type.value);
  } else {
    switch (type.type) {
      case 'Date':
        return Type.Referance('Date');
      case 'boolean':
        return Type.Boolean;
      case 'json':
        return Type.Unknown;
      case 'null':
        return Type.Null;
      case 'number':
        return Type.Number;
      case 'string':
        return Type.String;
      case 'unknown':
        return Type.Unknown;
    }
  }
};

export const toQueryTypescriptTypes = (context: DocumentContext, query: LoadedQuery) => {
  const paramsContext = withIdentifier(
    context,
    Type.Interface({
      name: 'QueryParams',
      isExport: true,
      props: query.params.map(({ name, type }) =>
        Type.Prop({ name, type: toPropertyType(type), isOptional: type.optional }),
      ),
    }),
  );

  const resultContext = withIdentifier(
    paramsContext,
    Type.Interface({
      name: 'QueryResult',
      isExport: true,
      props: query.result.map(({ name, type }) =>
        Type.Prop({ name, type: toPropertyType(type), isOptional: type.optional }),
      ),
    }),
  );

  return document(
    resultContext,
    Type.Interface({
      name: 'Query',
      isExport: true,
      props: [
        Type.Prop({ name: 'params', type: Type.Referance('QueryParams') }),
        Type.Prop({ name: 'result', type: Type.Referance('QueryResult') }),
      ],
    }),
  );
};

export const toQueryTypescript = (loadedQuery: LoadedQuery): string => {
  return printDocument(toQueryTypescriptTypes({}, loadedQuery));
};
