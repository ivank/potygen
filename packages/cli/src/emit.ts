import { mkdir, writeFile } from 'fs';
import { promisify } from 'util';
import { dirname, parse, relative } from 'path';
import {
  factory,
  createSourceFile,
  ScriptTarget,
  SourceFile,
  TypeNode,
  SyntaxKind,
  createPrinter,
  NewLineKind,
  Statement,
  Node,
  UnionTypeNode,
  PropertySignature,
  TypeLiteralNode,
  addSyntheticLeadingComment,
} from 'typescript';
import {
  LoadedFile,
  LoadedQueryInterface,
  isTypeArray,
  isTypeUnion,
  Type,
  isTypeObjectLiteral,
  isTypeLiteral,
  isTypeEqual,
  isTypeComposite,
  isUniqueBy,
  isTypeOptional,
  TypeName,
  isEmpty,
} from '@potygen/potygen';

const mkdirAsync = promisify(mkdir);
const writeFileAsync = promisify(writeFile);

const isIdentifierRegExp = /^[$A-Z_][0-9A-Z_$]*$/i;

const parseTemplate = (root: string, template: string, path: string): string =>
  Object.entries({ ...parse(relative(root, path)), root }).reduce(
    (acc, [name, value]) => acc.replace(`{{${name}}}`, value),
    template,
  );

type Refs = string[];

interface TypeContext {
  name: string;
  refs: Refs;
  toJson: boolean;
}

const jsDoc = (doc: string): string =>
  `*\n${doc
    .split('\n')
    .map((line) => ` * ${line}`)
    .join('\n')}\n `;

const withJSDoc = <T extends Node>(doc: string | undefined, node: T): T =>
  doc === undefined ? node : addSyntheticLeadingComment(node, SyntaxKind.MultiLineCommentTrivia, jsDoc(doc), true);

export const compactTypes = (types: Type[]): Type[] =>
  types
    .map((type) => (isTypeOptional(type) ? { ...type.value, nullable: type.nullable } : type))
    .filter((item, index, all) =>
      isTypeLiteral(item) && item.literal !== undefined
        ? !all.some(
            (other, otherIndex) => index !== otherIndex && other.type === item.type && other.literal === undefined,
          )
        : true,
    )
    .filter(isUniqueBy(isTypeEqual));

const toPropertyType =
  (context: TypeContext) =>
  (type: Type): TypeContext & { type: TypeNode } => {
    if (isTypeComposite(type)) {
      return { ...context, type: factory.createToken(SyntaxKind.StringKeyword) };
    } else if (isTypeObjectLiteral(type)) {
      return type.items.reduce<TypeContext & { type: TypeLiteralNode }>(
        (acc, item) => {
          const itemType = toPropertyType({ ...acc, name: context.name + toClassCase(item.name) })(item.type);
          const memeber = factory.createPropertySignature(
            undefined,
            item.name.includes(' ')
              ? factory.createComputedPropertyName(factory.createStringLiteral(item.name))
              : item.name,
            'nullable' in item.type && item.type.nullable ? factory.createToken(SyntaxKind.QuestionToken) : undefined,
            itemType.type,
          );
          return { ...itemType, type: factory.createTypeLiteralNode(acc.type.members.concat(memeber)) };
        },
        { ...context, type: factory.createTypeLiteralNode([]) },
      );
    } else if (isTypeArray(type)) {
      const itemsType = toPropertyType(context)(type.items);
      return { ...itemsType, type: factory.createArrayTypeNode(itemsType.type) };
    } else if (isTypeUnion(type)) {
      const unionTypes = compactTypes(type.items);
      return isEmpty(unionTypes)
        ? { ...context, type: factory.createToken(SyntaxKind.UnknownKeyword) }
        : unionTypes.reduce<TypeContext & { type: UnionTypeNode }>(
            (acc, item, index) => {
              const itemType = toPropertyType({ ...acc, name: context.name + index })(item);
              return { ...itemType, type: factory.createUnionTypeNode(acc.type.types.concat(itemType.type)) };
            },
            { ...context, type: factory.createUnionTypeNode([]) },
          );
    } else if (isTypeOptional(type)) {
      return type.value
        ? toPropertyType(context)(type.value)
        : { ...context, type: factory.createToken(SyntaxKind.UnknownKeyword) };
    } else {
      switch (type.type) {
        case TypeName.Date:
          return { ...context, type: factory.createTypeReferenceNode('Date') };
        case TypeName.Buffer:
          return { ...context, type: factory.createTypeReferenceNode('Buffer') };
        case TypeName.Boolean:
          return {
            ...context,
            type:
              type.literal !== undefined
                ? factory.createLiteralTypeNode(type.literal ? factory.createTrue() : factory.createFalse())
                : factory.createToken(SyntaxKind.BooleanKeyword),
          };
        case TypeName.Json:
          return {
            ...context,
            refs: context.refs.concat(context.name),
            type: context.toJson
              ? factory.createTypeReferenceNode('Json', [factory.createTypeReferenceNode(context.name)])
              : factory.createTypeReferenceNode(context.name),
          };
        case TypeName.Null:
          return { ...context, type: factory.createLiteralTypeNode(factory.createToken(SyntaxKind.NullKeyword)) };
        case TypeName.Number:
        case TypeName.BigInt:
          return {
            ...context,
            type:
              type.literal !== undefined
                ? factory.createLiteralTypeNode(factory.createStringLiteral(String(type.literal)))
                : factory.createToken(SyntaxKind.NumberKeyword),
          };
        case TypeName.String:
          return {
            ...context,
            type:
              type.literal !== undefined
                ? factory.createLiteralTypeNode(factory.createStringLiteral(type.literal))
                : factory.createToken(SyntaxKind.StringKeyword),
          };
        case TypeName.Unknown:
          return { ...context, type: factory.createToken(SyntaxKind.UnknownKeyword) };
        case TypeName.Any:
          return { ...context, type: factory.createToken(SyntaxKind.AnyKeyword) };
      }
    }
  };

const toClassCase = (identifier: string) => identifier[0].toUpperCase() + identifier.slice(1);

const toAstImports = (names: string[]): Statement =>
  factory.createImportDeclaration(
    undefined,
    undefined,
    factory.createImportClause(
      false,
      undefined,
      factory.createNamedImports(
        names.map((name) => factory.createImportSpecifier(false, undefined, factory.createIdentifier(name))),
      ),
    ),
    factory.createStringLiteral('@potygen/potygen'),
  );

const toLoadedQueryTypeNodes = (
  refs: Refs,
  name: string,
  loadedQuery: LoadedQueryInterface,
): { resultRefs: Refs; statements: Statement[] } => {
  const params = loadedQuery.params.reduce<{ refs: Refs; props: PropertySignature[] }>(
    (acc, item) => {
      const itemType = toPropertyType({ toJson: false, name: 'TParam' + toClassCase(item.name), refs: acc.refs })(
        item.type,
      );
      const prop = factory.createPropertySignature(
        undefined,
        item.name,
        'nullable' in item.type && item.type.nullable ? factory.createToken(SyntaxKind.QuestionToken) : undefined,
        itemType.type,
      );
      return { ...itemType, props: acc.props.concat(prop) };
    },
    { refs, props: [] },
  );

  const results = loadedQuery.results.reduce<{ refs: Refs; props: PropertySignature[] }>(
    (acc, item) => {
      const itemType = toPropertyType({ toJson: true, name: 'T' + toClassCase(item.name), refs: acc.refs })(item.type);
      const prop = withJSDoc(
        item.type.comment ?? undefined,
        factory.createPropertySignature(
          undefined,
          isIdentifierRegExp.test(item.name) ? item.name : factory.createStringLiteral(item.name),
          'nullable' in item.type && item.type.nullable ? factory.createToken(SyntaxKind.QuestionToken) : undefined,
          itemType.type,
        ),
      );
      return { ...itemType, props: acc.props.concat(prop) };
    },
    { refs: [], props: [] },
  );

  return {
    resultRefs: results.refs,
    statements: [
      factory.createInterfaceDeclaration(
        undefined,
        [factory.createModifier(SyntaxKind.ExportKeyword)],
        `${name}Params`,
        params.refs.map((ref) =>
          factory.createTypeParameterDeclaration(ref, undefined, factory.createToken(SyntaxKind.UnknownKeyword)),
        ),
        undefined,
        params.props,
      ),
      factory.createInterfaceDeclaration(
        undefined,
        [factory.createModifier(SyntaxKind.ExportKeyword)],
        `${name}Result`,
        results.refs.map((ref) =>
          factory.createTypeParameterDeclaration(ref, undefined, factory.createToken(SyntaxKind.UnknownKeyword)),
        ),
        undefined,
        results.props,
      ),
      factory.createInterfaceDeclaration(
        undefined,
        [factory.createModifier(SyntaxKind.ExportKeyword)],
        `${name}Query`,
        [...params.refs, ...results.refs].map((ref) =>
          factory.createTypeParameterDeclaration(ref, undefined, factory.createToken(SyntaxKind.UnknownKeyword)),
        ),
        undefined,
        [
          factory.createPropertySignature(
            undefined,
            'params',
            undefined,
            factory.createTypeReferenceNode(
              `${name}Params`,
              params.refs.map((ref) => factory.createTypeReferenceNode(ref)),
            ),
          ),
          factory.createPropertySignature(
            undefined,
            'result',
            undefined,
            factory.createTypeReferenceNode(
              `${name}Result`,
              results.refs.map((ref) => factory.createTypeReferenceNode(ref)),
            ),
          ),
        ],
      ),
    ],
  };
};

export const toTypeSource = (file: LoadedFile, typePrefix: string = ''): SourceFile => {
  const content =
    file.type === 'ts'
      ? file.queries.reduce<{ resultRefs: Refs; statements: Statement[] }>(
          (acc, query) => {
            const { statements, resultRefs } = toLoadedQueryTypeNodes(
              acc.resultRefs,
              typePrefix + toClassCase(query.name),
              query.loadedQuery,
            );
            return { resultRefs, statements: acc.statements.concat(statements) };
          },
          { resultRefs: [], statements: [] },
        )
      : toLoadedQueryTypeNodes([], typePrefix, file.loadedQuery);

  return factory.updateSourceFile(
    createSourceFile(file.path, '', ScriptTarget.ES2021, true),
    content.resultRefs.length > 0 ? [toAstImports(['Json']), ...content.statements] : content.statements,
  );
};

export const emitLoadedFile = (root: string, template: string, typePrefix?: string) => {
  const printer = createPrinter({ newLine: NewLineKind.LineFeed });
  return async (file: LoadedFile): Promise<void> => {
    const outputFile = parseTemplate(root, template, file.path);
    const directory = dirname(outputFile);

    await mkdirAsync(directory, { recursive: true });
    await writeFileAsync(outputFile, printer.printFile(toTypeSource(file, typePrefix)));
  };
};
