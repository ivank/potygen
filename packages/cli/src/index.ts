export {
  DataTable,
  DataFunction,
  DataEnum,
  DataView,
  DataComposite,
  LoadedDataTable,
  LoadedDataComposite,
  DataViewRaw,
  DataViewParsed,
  LoadedDataView,
  LoadedDataFunction,
  LoadedDataEnum,
  Data,
  LoadedData,
  LoadedParam,
  LoadedResult,
  LoadedQueryInterface,
  LoadedFunction,
  LoadedComposite,
  LoadedSourceTable,
  LoadedSourceQuery,
  LoadedSourceView,
  LoadedSourceValues,
  LoadedSource,
  LoadedContext,
  TemplateTagQuery,
  ParsedTypescriptFile,
  ParsedSqlFile,
  ParsedFile,
  LoadedTemplateTagQuery,
  LoadedTypescriptFile,
  LoadedSqlFile,
  LoadedFile,
  Logger,
  LoadContext,
  QualifiedName,
} from './types';

export {
  isDataTable,
  isDataFunction,
  isDataEnum,
  isLoadedDataTable,
  isLoadedDataFunction,
  isLoadedDataEnum,
  isLoadedDataView,
  isLoadedDataComposite,
} from './guards';

export { toLoadedContext, loadQueryInterfacesData, toLoadedQueryInterface, loadData, extractDataSources } from './load';

export { Config, ConfigType, FullConfigType, toConfig } from './config';

export { emitLoadedFile, toTypeSource } from './emit';
export { glob } from './glob';
