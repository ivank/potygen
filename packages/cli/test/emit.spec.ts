import { Client } from 'pg';
import {
  parser,
  toQueryInterface,
  loadQueryInterfacesData,
  toLoadedQueryInterface,
  TypeConstant,
  TypeName,
} from '@potygen/potygen';
import { createPrinter, NewLineKind } from 'typescript';
import { toTypeSource, compactTypes } from '../src';
import { testDb } from './helpers';

let db: Client;

describe('Query Interface', () => {
  beforeAll(async () => {
    db = testDb();
    await db.connect();
  });

  afterAll(() => db.end());

  it.each<[string, string]>([
    ['function result single', `SELECT ABS(integer_col) FROM all_types`],
    ['function result double', `SELECT ABS(ABS(integer_col)) FROM all_types`],
    ['load nullable column', `SELECT integer_col FROM all_types`],
    ['load non nullable column', `SELECT not_null FROM all_types`],
    ['load nullable column', `SELECT integer_col::text FROM all_types`],
    ['load non nullable column', `SELECT not_null::text FROM all_types`],
    ['nullability', `SELECT not_null FROM all_types`],
    ['nested function guess type', `SELECT ABS(ARRAY_LENGTH(ARRAY_AGG(integer_col), 1)) FROM all_types GROUP BY id`],
    ['nested function explicit type', `SELECT (ARRAY_AGG(integer_col), 1)::int[] FROM all_types GROUP BY id`],
    ['operators integer', `SELECT integer_col + integer_col AS "test1" FROM all_types WHERE id = $id`],
    ['operators string', `SELECT character_col + integer_col AS "test1" FROM all_types WHERE character_col = $text`],
    ['different result types', `SELECT * FROM all_types`],
    ['enum', `SELECT 'Pending'::account_levelisation_state`],
    ['enum column', `SELECT state FROM account_levelisations`],
    ['simple', `SELECT id, character_col FROM all_types WHERE id = :id`],
    [
      'insert',
      `INSERT INTO all_types(not_null, integer_col, character_col) VALUES $$vals(notNull, integerCol, characterCol)`,
    ],
    ['parameter group collapse', `SELECT id FROM all_types WHERE $q = '' OR character_varying_col = $q`],
  ])('Should convert %s sql (%s)', async (path, content) => {
    const logger = { info: jest.fn(), error: jest.fn(), debug: jest.fn() };
    const printer = createPrinter({ newLine: NewLineKind.LineFeed });
    const { ast } = parser(content);
    const queryInterface = toQueryInterface(ast);

    const data = await loadQueryInterfacesData({ db, logger }, [queryInterface], []);
    const loadedQuery = toLoadedQueryInterface(data)(queryInterface);
    const source = toTypeSource({ type: 'sql', path, content, queryInterface, loadedQuery });
    expect(printer.printFile(source)).toMatchSnapshot();
  });

  it.each<[string, TypeConstant[], TypeConstant[]]>([
    [
      'single value literal',
      [
        { type: TypeName.Boolean, literal: true, postgresType: 'boolean' },
        { type: TypeName.Boolean, postgresType: 'boolean' },
      ],
      [{ type: TypeName.Boolean, postgresType: 'boolean' }],
    ],
    [
      'keep if no need to compact',
      [
        { type: TypeName.String, literal: 'tmp', postgresType: 'text' },
        { type: TypeName.String, literal: 'tmp2', postgresType: 'text' },
      ],
      [
        { type: TypeName.String, literal: 'tmp', postgresType: 'text' },
        { type: TypeName.String, literal: 'tmp2', postgresType: 'text' },
      ],
    ],
    [
      'compact if at least one non literal',
      [
        { type: TypeName.String, literal: 'tmp', postgresType: 'text' },
        { type: TypeName.String, literal: 'tmp2', postgresType: 'text' },
        { type: TypeName.String, postgresType: 'text' },
      ],
      [{ type: TypeName.String, postgresType: 'text' }],
    ],
  ])('Should compact union types for %s', async (_, types, expected) => {
    expect(compactTypes(types)).toEqual(expected);
  });
});
