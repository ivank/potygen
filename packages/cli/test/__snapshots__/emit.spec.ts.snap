// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query Interface Should convert array constructor sql (SELECT ARRAY(SELECT json_build_object('test', other.integer_col) FROM all_types AS other WHERE all_types.id = other.id) as "arr" FROM all_types) 1`] = `
"export interface Params {
}
export interface Result {
    arr: {
        test?: number;
    }[];
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert case to string literal union sql (SELECT CASE accounts.source_system WHEN 'SSE' THEN 'SSE' WHEN 'FITDB' THEN 'FITDB' ELSE NULL END AS "sourceSystem" FROM accounts) 1`] = `
"export interface Params {
}
export interface Result {
    sourceSystem: \\"SSE\\" | \\"FITDB\\" | undefined;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert descriptive property names sql (SELECT 'test' AS "some column", 'test2' AS "test") 1`] = `
"export interface Params {
}
export interface Result {
    \\"some column\\": \\"test\\";
    test: \\"test2\\";
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert different result types for returning sql (UPDATE all_types SET id = 1 RETURNING *) 1`] = `
"import { Json } from \\"@potygen/potygen\\";
export interface Params {
}
export interface Result<TJson_col = unknown, TJsonb_col = unknown> {
    id: number;
    /**
     * This column should never be null
     */
    not_null: number;
    default_not_null: number;
    bigint_col?: number;
    bigserial_col: number;
    bit_col?: string;
    bit_varying_col?: string;
    boolean_col?: boolean;
    box_col?: string;
    bytea_col: Buffer;
    character_col?: string;
    character_varying_col?: string;
    cidr_col?: string;
    circle_col: {
        x: number;
        y: number;
        radius: number;
    };
    date_col?: Date;
    double_col?: number;
    inet_col?: string;
    integer_col?: number;
    interval_col: {
        years: number;
        months: number;
        days: number;
        hours: number;
        minutes: number;
        seconds: number;
        milliseconds: number;
    };
    json_col?: Json<TJson_col>;
    jsonb_col?: Json<TJsonb_col>;
    line_col?: string;
    lseg_col?: string;
    macaddr_col?: string;
    money_col?: string;
    numeric_col?: number;
    path_col?: string;
    pg_lsn_col?: string;
    point_col: {
        x: number;
        y: number;
    };
    polygon_col?: string;
    real_col?: number;
    smallint_col?: number;
    smallserial_col: number;
    serial_col: number;
    text_col?: string;
    time_col?: string;
    time_with_time_zone_col?: Date;
    timestamp_col?: Date;
    timestamp_with_time_zone_col?: Date;
    tsquery_col?: string;
    tsvector_col?: string;
    txid_snapshot_col?: string;
    uuid_col?: string;
    xml_col?: string;
    item: string;
    /**
     *
     *   This column is an enum.
     *   It also contains several lines:
     *     - line1
     *     - line2
     *
     */
    state?: \\"Active\\" | \\"Pending\\" | \\"Dispute\\" | \\"Closed\\";
    state_arr?: (\\"Active\\" | \\"Pending\\" | \\"Dispute\\" | \\"Closed\\")[];
    item_arr?: string[];
    static_arr?: number[];
}
export interface Query<TJson_col = unknown, TJsonb_col = unknown> {
    params: Params;
    result: Result<TJson_col, TJsonb_col>;
}
"
`;

exports[`Query Interface Should convert different result types sql (SELECT * FROM all_types) 1`] = `
"import { Json } from \\"@potygen/potygen\\";
export interface Params {
}
export interface Result<TJson_col = unknown, TJsonb_col = unknown> {
    id: number;
    /**
     * This column should never be null
     */
    not_null: number;
    default_not_null: number;
    bigint_col?: number;
    bigserial_col: number;
    bit_col?: string;
    bit_varying_col?: string;
    boolean_col?: boolean;
    box_col?: string;
    bytea_col: Buffer;
    character_col?: string;
    character_varying_col?: string;
    cidr_col?: string;
    circle_col: {
        x: number;
        y: number;
        radius: number;
    };
    date_col?: Date;
    double_col?: number;
    inet_col?: string;
    integer_col?: number;
    interval_col: {
        years: number;
        months: number;
        days: number;
        hours: number;
        minutes: number;
        seconds: number;
        milliseconds: number;
    };
    json_col?: Json<TJson_col>;
    jsonb_col?: Json<TJsonb_col>;
    line_col?: string;
    lseg_col?: string;
    macaddr_col?: string;
    money_col?: string;
    numeric_col?: number;
    path_col?: string;
    pg_lsn_col?: string;
    point_col: {
        x: number;
        y: number;
    };
    polygon_col?: string;
    real_col?: number;
    smallint_col?: number;
    smallserial_col: number;
    serial_col: number;
    text_col?: string;
    time_col?: string;
    time_with_time_zone_col?: Date;
    timestamp_col?: Date;
    timestamp_with_time_zone_col?: Date;
    tsquery_col?: string;
    tsvector_col?: string;
    txid_snapshot_col?: string;
    uuid_col?: string;
    xml_col?: string;
    item: string;
    /**
     *
     *   This column is an enum.
     *   It also contains several lines:
     *     - line1
     *     - line2
     *
     */
    state?: \\"Active\\" | \\"Pending\\" | \\"Dispute\\" | \\"Closed\\";
    state_arr?: (\\"Active\\" | \\"Pending\\" | \\"Dispute\\" | \\"Closed\\")[];
    item_arr?: string[];
    static_arr?: number[];
}
export interface Query<TJson_col = unknown, TJsonb_col = unknown> {
    params: Params;
    result: Result<TJson_col, TJsonb_col>;
}
"
`;

exports[`Query Interface Should convert empty array sql (SELECT ARRAY[]) 1`] = `
"export interface Params {
}
export interface Result {
    array: unknown[];
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert enum column sql (SELECT state FROM account_levelisations) 1`] = `
"export interface Params {
}
export interface Result {
    state: \\"Pending\\" | \\"Done\\";
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert enum sql (SELECT 'Pending'::account_levelisation_state) 1`] = `
"export interface Params {
}
export interface Result {
    row: \\"Pending\\" | \\"Done\\";
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert function result double sql (SELECT ABS(ABS(integer_col)) FROM all_types) 1`] = `
"export interface Params {
}
export interface Result {
    abs: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert function result single sql (SELECT ABS(integer_col) FROM all_types) 1`] = `
"export interface Params {
}
export interface Result {
    abs: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert insert sql (INSERT INTO all_types(not_null, integer_col, character_col) VALUES $$vals(notNull, integerCol, characterCol)) 1`] = `
"export interface Params {
    vals: {
        notNull?: number;
        integerCol?: number;
        characterCol?: string;
    }[];
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert invalid identifiers as property names sql (SELECT 'test' AS "12") 1`] = `
"export interface Params {
}
export interface Result {
    \\"12\\": \\"test\\";
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert json returning sql (UPDATE all_types SET id = 1 RETURNING json_col) 1`] = `
"import { Json } from \\"@potygen/potygen\\";
export interface Params {
}
export interface Result<TJson_col = unknown> {
    json_col?: Json<TJson_col>;
}
export interface Query<TJson_col = unknown> {
    params: Params;
    result: Result<TJson_col>;
}
"
`;

exports[`Query Interface Should convert load non nullable column sql (SELECT not_null FROM all_types) 1`] = `
"export interface Params {
}
export interface Result {
    bool: boolean;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert load non nullable column sql (SELECT not_null::text FROM all_types) 1`] = `
"export interface Params {
}
export interface Result {
    bool: boolean;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert load nullable column sql (SELECT integer_col FROM all_types) 1`] = `
"export interface Params {
}
export interface Result {
    integer_col?: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert load nullable column sql (SELECT integer_col::text FROM all_types) 1`] = `
"export interface Params {
}
export interface Result {
    integer_col?: string;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert nested function explicit type sql (SELECT (ARRAY_AGG(integer_col), 1)::int[] FROM all_types GROUP BY id) 1`] = `
"export interface Params {
}
export interface Result {
    array: number[];
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert nested function guess type sql (SELECT ABS(ARRAY_LENGTH(ARRAY_AGG(integer_col), 1)) FROM all_types GROUP BY id) 1`] = `
"export interface Params {
}
export interface Result {
    abs: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert nullability sql (SELECT not_null FROM all_types) 1`] = `
"export interface Params {
}
export interface Result {
    bool: boolean;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert operators integer sql (SELECT integer_col + integer_col AS "test1" FROM all_types WHERE id = $id) 1`] = `
"export interface Params {
    id?: number;
}
export interface Result {
    test1?: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert operators string sql (SELECT character_col + integer_col AS "test1" FROM all_types WHERE character_col = $text) 1`] = `
"export interface Params {
    text?: string;
}
export interface Result {
    test1?: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert parameter group collapse sql (SELECT id FROM all_types WHERE $q = '' OR character_varying_col = $q) 1`] = `
"export interface Params {
    q?: string;
}
export interface Result {
    id: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Query Interface Should convert simple sql (SELECT id, character_col FROM all_types WHERE id = :id) 1`] = `
"export interface Params {
    id?: number;
}
export interface Result {
    id: number;
    character_col?: string;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;
